cmd/snap-bootstrap/blkid/blkid.go

```diff
@@ -0,0 +1,171 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package blkid
+
+//#cgo CFLAGS: -D_FILE_OFFSET_BITS=64
+//#cgo pkg-config: blkid
+//#cgo LDFLAGS:
+//
+//#include <stdlib.h>
+//#include <blkid.h>
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+const (
+	BLKID_PARTS_ENTRY_DETAILS int = C.BLKID_PARTS_ENTRY_DETAILS
+)
+
+
+// AbstractBlkidProbe is wrapper for blkid_probe
+// See "Low-level" section of libblkid documentation
+type AbstractBlkidProbe interface {
+	// LookupValue is a wrapper for blkid_probe_lookup_value
+	LookupValue(entryName string) (string, error)
+	// Close is a wrapper for blkid_free_probe
+	Close()
+	// EnablePartitions is a wrapper for blkid_probe_enable_partitions
+	EnablePartitions(value bool)
+	// EnableSuperblocks is a wrapper for blkid_probe_enable_superblocks
+	EnableSuperblocks(value bool)
+	// SetPartitionsFlags is a wrapper for blkid_probe_set_partitions_flags
+	SetPartitionsFlags(flags int)
+	// DoSafeprobe is a wrapper for blkid_do_safeprobe
+	DoSafeprobe() error
+	// GetPartitions is a wrapper for blkid_probe_get_partitions
+	GetPartitions() (AbstractBlkidPartlist, error)
+}
+
+// AbstractBlkidPartlist is a wrapper for blkid_partlist
+type AbstractBlkidPartlist interface {
+	// GetPartitions is a wrapper for blkid_partlist_get_partition
+	// and blkid_partlist_numof_partitions.
+	GetPartitions() []AbstractBlkidPartition
+}
+
+// AbstractBlkidPartition is a wrapper for blkid_partition
+type AbstractBlkidPartition interface {
+	// GetName is a wrapper for blkid_partition_get_name
+	GetName() string
+	// GetUUID is a wrapper for blkid_partition_get_uuid
+	GetUUID() string
+}
+
+type blkidProbe struct {
+	probeHandle C.blkid_probe
+}
+
+func newProbeFromFilenameImpl(node string) (AbstractBlkidProbe, error) {
+	cnode := C.CString(node)
+	defer C.free(unsafe.Pointer(cnode))
+	C.blkid_new_probe_from_filename(cnode)
+	probe, err := C.blkid_new_probe_from_filename(cnode)
+	if probe == nil {
+		return nil, err
+	}
+	return &blkidProbe{probe}, nil
+}
+
+var NewProbeFromFilename = newProbeFromFilenameImpl
+
+func (p *blkidProbe) LookupValue(entryName string) (string, error) {
+	var value *C.char
+	var value_len C.size_t
+	cname := C.CString(entryName)
+	defer C.free(unsafe.Pointer(cname))
+	res := C.blkid_probe_lookup_value(p.probeHandle, cname, &value, &value_len)
+	if res < 0 {
+		return "", fmt.Errorf("Probe value was not found: %s", entryName)
+	}
+	if value_len > 0 {
+		return C.GoStringN(value, C.int(value_len-1)), nil
+	} else {
+		return "", fmt.Errorf("Probe value has unexpected size")
+	}
+}
+
+func (p *blkidProbe) Close() {
+	C.blkid_free_probe(p.probeHandle)
```

Comment:
should we put a nil value into p.probeHandle after the free? and check it in the caller? as things is it wouldn't be impossible to try to use the probe after is is closed

Reply:
Done